odl/solvers/functional/default_functionals.py

```diff
@@ -0,0 +1,628 @@
+# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Default functionals defined on any (reasonable) space."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+from builtins import super
+
+import numpy as np
+
+from odl.solvers.functional.functional import Functional
+from odl.operator.operator import Operator
+from odl.solvers.advanced.proximal_operators import (proximal_l1,
+                                                     proximal_cconj_l1,
+                                                     proximal_l2,
+                                                     proximal_cconj_l2,
+                                                     proximal_l2_squared)
+
+from odl.operator.default_ops import (ZeroOperator, IdentityOperator)
+
+
+__all__ = ('L1Norm', 'L2Norm', 'L2NormSquare', 'ZeroFunctional',
+           'ConstantFunctional')
+
+
+# TODO: Implement some of the missing gradients
+
+class L1Norm(Functional):
+    """The functional corresponding to L1-norm."""
+
+    def __init__(self, domain):
+        """Initalizes an instance of the L1-norm functional.
+
+        Parameters
+        ----------
+        domain : `LinearSpace`
+            Domain of the functional.
+        """
+        super().__init__(domain=domain, linear=False, convex=True,
+                         concave=False, smooth=False, grad_lipschitz=np.inf)
+
+    def _call(self, x):
+        """Applies the functional to the given point.
+
+        Parameters
+        ----------
+        x : `LinearSpaceVector`
+            Element in the domain of the functional.
+
+        Returns
+        -------
+        `self(x)` : `element` in the `field` of the ``domain``.
+            Evaluation of the functional.
+        """
+        return np.abs(x).inner(self.domain.one())
+
+    @property
+    def gradient(x):
+        """Gradient operator of the L1-functional."""
+        # It does not exist on all of the space
+        raise NotImplementedError
```

Comment:
Perhaps with a message, `raise NotImplementedError('functional not differentiable')`.